bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java

```diff
@@ -883,12 +884,51 @@ public void process(int journalVersion, long offset, ByteBuffer recBuff) throws
         };
 
         for (Journal journal : journals) {
-            journal.replay(scanner);
+            replay(journal, scanner);
         }
         long elapsedTs = System.currentTimeMillis() - startTs;
         LOG.info("Finished replaying journal in {} ms.", elapsedTs);
     }
 
+    /**
+     * Replay journal files and updates journal's in-memory lastLogMark object.
+     *
+     * @param journal Journal object corresponding to a journalDir
+     * @param scanner Scanner to process replayed entries.
+     * @throws IOException
+     */
+    private void replay(Journal journal, JournalScanner scanner) throws IOException {
+        final LogMark markedLog = journal.getLastLogMark().getCurMark();
+        List<Long> logs = Journal.listJournalIds(journal.getJournalDirectory(), journalId -> {
+            if (journalId < markedLog.getLogFileId()) {
+                return false;
+            }
+            return true;
+        });
+        // last log mark may be missed due to no sync up before
+        // validate filtered log ids only when we have markedLogId
+        if (markedLog.getLogFileId() > 0) {
+            if (logs.size() == 0 || logs.get(0) != markedLog.getLogFileId()) {
+                throw new IOException("Recovery log " + markedLog.getLogFileId() + " is missing");
+            }
+        }
+
+        // TODO: When reading in the journal logs that need to be synced, we
+        // should use BufferedChannels instead to minimize the amount of
+        // system calls done.
+        for (Long id : logs) {
+            long logPosition = 0L;
+            if (id == markedLog.getLogFileId()) {
+                logPosition = markedLog.getLogFileOffset();
+            }
+            LOG.info("Replaying journal {} from position {}", id, logPosition);
+            journal.scanJournal(id, logPosition, scanner);
+            // Update LastLogMark to Long.MAX_VALUE position after replaying journal
+            // After LedgerStorage flush, SyncThread should persist this to disk
+            journal.getLastLogMark().setCurLogMark(id, Long.MAX_VALUE);
```

Comment:
I missed noticing this, I'm not sure if it is completely ok to store Long.MAX_VALUE as its position.

Reply:
I referred this: https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileChannel.html#position(long)

Since we don't have a clear way to know the exact EOF, this is a work around for that.

Reply:
there are multiple things which we have to make sure about this
1) under no circumstance, we would try to write to this journal file beyond this curLogMark (Long.MAX_VALUE position).
2) What if a reader tries to read this journal file upto this Long.MAX_VALUE position? say for eg - readJournal BookieShell command.

In summary we have to be sure that all the components in our code deals correctly with this special position.

Reply:
> under no circumstance, we would try to write to this journal file beyond this curLogMark (Long.MAX_VALUE position).

Bookie restarts always create a new journal file with default size of 16 MB. This ensures that we never write to old journal file. This scenario is covered by the test case.

> What if a reader tries to read this journal file upto this Long.MAX_VALUE position? say for eg - readJournal BookieShell command

Some text from the link in my previous comment, `Setting the position to a value that is greater than the file's current size is legal but does not change the size of the file. A later attempt to read bytes at such a position will immediately return an end-of-file indication`, should help resolve it.

I understand that this is a critical fix and we have to think through all possibilities. IMHO, the test case (although a bit long), tries to cover all such potential scenarios.

Reply:
@sijie @jvrao I don't have strong case to say no to this change, but I would like to get second opinion on this. Since it falls under critical section, I just want to double confirm.

Reply:
I have fixed the required tests and checkstyle violations. Other tests are failing b/c they are flaky. I have verified `testEmptyLedgerLosesQuorumEventually` by running locally.

Reply:
@karanmehta93 i'm approving this PR. @sijie please let us know if we missed anything with respect to setting CurLogMark to Long.MAX_VALUE.